### PR TITLE
KAA-825: Add topic list hash

### PR DESCRIPTION
**Pull request by ohamov**

### CODE REVIEW

File: client/client-multi/client-cpp/impl/ClientStatus.cpp [hunk 1]
```diff
@@ -529,6 +534,14 @@ void ClientStatus::read()
     std::string token;
     std::stringstream stream;
 
+	/* First read topicList hash */
+    if (stateFile.good()) {
+	    std::getline(stateFile, token, '=');
```
Comment (ghost) [inline client/client-multi/client-cpp/impl/ClientStatus.cpp]: Fix indentation.

File: client/client-multi/client-cpp/impl/notification/NotificationTransport.cpp [hunk 2]
```diff
@@ -114,42 +105,64 @@ NotificationSyncRequestPtr NotificationTransport::createNotificationRequest()
...
+                 tId = topic.id;
+                 topicListHash = 31 * topicListHash + (uint32_t)(tId ^ (tId >> 32));
+            }
+            if (clientStatus_->getTopicListHash() != topicListHash) {
+                clientStatus_->setTopicListHash(topicListHash);
+
+                /* Add mandatory topics to topic state list */
+                for (const auto& topic : topics) {
+                if (topic.subscriptionType == MANDATORY_SUBSCRIPTION) {
+                DetailedTopicState dts = { topic.id, topic.name, topic.subscriptionType, 0 };
```
Comment (ghost) [inline client/client-multi/client-cpp/impl/notification/NotificationTransport.cpp]: Fix indentation.

File: client/client-multi/client-cpp/kaa/logging/LoggingUtils.hpp [hunk 3]
```diff
@@ -54,9 +54,11 @@ class LoggingUtils {
     static std::string ConfigurationSyncRequestToString(const SyncRequest::configurationSyncRequest_t& request) {
         std::ostringstream ss;
         if (!request.is_null()) {
-            ss << KVSTRING(appStateSeqNumber, request.get_ConfigurationSyncRequest().appStateSeqNumber);
-            ss << KVSTRING(configurationHash,
-                    (request.get_ConfigurationSyncRequest().configurationHash.is_null() ? "null" : ByteArrayToString(request.get_ConfigurationSyncRequest().configurationHash.get_bytes())));
+		  std::stringstream hashStream;
+		  for (auto &byte : request.get_ConfigurationSyncRequest().configurationHash) {
+               hashStream << byte;
```
Comment (ghost) [inline client/client-multi/client-cpp/kaa/logging/LoggingUtils.hpp]: Fix indentation

File: client/client-multi/client-cpp/impl/ClientStatus.cpp [hunk 4]
```diff
@@ -62,9 +62,9 @@ static bimap create_bimap()
     return bi;
 }
 
-const bimap                 ClientStatus::parameterToToken_ = create_bimap();
-const SequenceNumber        ClientStatus::appSeqNumberDefault_ = {0, 0, 0};
-const bool                  ClientStatus::isRegisteredDefault_ = false;
+const bimap                 ClientStatus::parameterToToken_ =   create_bimap();
```
Comment (forGGe) [inline client/client-multi/client-cpp/impl/ClientStatus.cpp:65]: Odd whitespace

File: client/client-multi/client-cpp/impl/KaaClient.cpp [hunk 5]
```diff
@@ -388,15 +388,15 @@ void KaaClient::removeNotificationListener(INotificationListener& listener) {
 #endif
 }
 
-void KaaClient::removeNotificationListener(const std::string& topidId, INotificationListener& listener) {
+  void KaaClient::removeNotificationListener(std::int64_t topicId, INotificationListener& listener) {
```
Comment (forGGe) [inline client/client-multi/client-cpp/impl/KaaClient.cpp]: odd whitespace

File: client/client-multi/client-cpp/kaa/ClientStatus.hpp [hunk 6]
```diff
@@ -100,6 +98,7 @@ class ClientStatus : public IKaaClientStateStorage {
     bool hasUpdate_;
 
     KAA_MUTEX_MUTABLE_DECLARE(sequenceNumberGuard_);
+    std::uint32_t topicListHash_;
 
     static const bimap                      parameterToToken_;
     static const SequenceNumber             appSeqNumberDefault_;
```
Comment (forGGe) [inline client/client-multi/client-cpp/kaa/ClientStatus.hpp:107]: Per discussion: code that operates over this variable must be altered

File: client/client-multi/client-cpp/test/impl/ClientStatusTest.cpp [hunk 7]
```diff
@@ -77,13 +73,15 @@ BOOST_AUTO_TEST_CASE(checkSetAndSaveParameters)
 
     DetailedTopicStates ts;
     DetailedTopicState ts1;
-    ts1.topicId = "topic1";
+	std::int64_t topic1 = 0x01;
```
Comment (forGGe) [inline client/client-multi/client-cpp/test/impl/ClientStatusTest.cpp]: Fix indentation

**Comment by forGGe** [comment]
Reviewed.

File: client/client-multi/client-cpp/impl/ClientStatus.cpp (hunk 1 above)
Comment (ohamov) [inline client/client-multi/client-cpp/impl/ClientStatus.cpp]: Done

File: client/client-multi/client-cpp/impl/KaaClient.cpp (hunk 5 above)
Comment (ohamov) [inline client/client-multi/client-cpp/impl/KaaClient.cpp]: Fixed

File: client/client-multi/client-cpp/impl/notification/NotificationTransport.cpp (hunk 2 above)
Comment (ohamov) [inline client/client-multi/client-cpp/impl/notification/NotificationTransport.cpp]: Done

File: client/client-multi/client-cpp/kaa/ClientStatus.hpp (hunk 6 above)
Comment (ohamov) [inline client/client-multi/client-cpp/kaa/ClientStatus.hpp:107]: Done

File: client/client-multi/client-cpp/kaa/logging/LoggingUtils.hpp (hunk 3 above)
Comment (ohamov) [inline client/client-multi/client-cpp/kaa/logging/LoggingUtils.hpp]: Done

File: client/client-multi/client-cpp/test/impl/ClientStatusTest.cpp (hunk 7 above)
Comment (ohamov) [inline client/client-multi/client-cpp/test/impl/ClientStatusTest.cpp]: Fixed

File: client/client-multi/client-cpp/impl/ClientStatus.cpp (hunk 4 above)
Comment (ohamov) [inline client/client-multi/client-cpp/impl/ClientStatus.cpp:65]: Take a look what indent have the variables which are below

File: client/client-multi/client-cpp/impl/notification/NotificationTransport.cpp [hunk 8]
```diff
@@ -29,33 +29,25 @@
...
-    /*TODO: topic list hash is a future feature */
-    request->topicListHash.set_null();
+    request->topicListHash = clientStatus_->getTopicListHash();
 
-    const DetailedTopicStates& detailedStatesContainer = clientStatus_->getTopicStates();
-    if (!detailedStatesContainer.empty()) {
-        std::vector<TopicState> container(detailedStatesContainer.size());
-        auto it = detailedStatesContainer.begin();
+	const auto& topicStates = clientStatus_->getTopicStates();
+    if (!topicStates.empty()) {
```
Comment (ghost) [inline client/client-multi/client-cpp/impl/notification/NotificationTransport.cpp:57]: This code snippet duplicates both in createNotificationRequest() and in createEmptyNotificationRequest().
Could you move it to a separate function?

Comment (ohamov) [inline client/client-multi/client-cpp/impl/notification/NotificationTransport.cpp:57]: Done

File: client/client-multi/client-cpp/impl/ClientStatus.cpp (hunk 4 above)
Comment (forGGe) [inline client/client-multi/client-cpp/impl/ClientStatus.cpp:65]: Good

**Comment by forGGe** [comment]
LGTM

**Comment by ghost** [comment]
LGTM

**Pull request merged**